src/lib/comms/sol-http-common.c

```diff
@@ -74,52 +130,645 @@ sol_http_param_free(struct sol_http_param *params)
+            params);
+        SOL_INT_CHECK_GOTO(r, < 0, exit);
+    }
+
+    if (url.fragment.len) {
+        r = sol_buffer_append_char(&buf, '#');
+        SOL_INT_CHECK_GOTO(r, < 0, exit);
+        r = sol_buffer_append_slice(&buf, url.fragment);
+        SOL_INT_CHECK_GOTO(r, < 0, exit);
     }
 
     r = sol_buffer_ensure_nul_byte(&buf);
-    SOL_INT_CHECK_GOTO(r, < 0, end);
+    SOL_INT_CHECK_GOTO(r, < 0, exit);
+
+    *url_out = sol_buffer_steal(&buf, NULL);
+exit:
+    sol_buffer_fini(&buf);
+    return r;
+}
+
+SOL_API int
+sol_http_create_simple_uri(char **uri, const struct sol_str_slice base_url,
+    const struct sol_http_param *params)
+{
+    struct sol_buffer buf;
+    int r;
+
+    SOL_NULL_CHECK(uri, -EINVAL);
+    if (base_url.len == 0) {
+        SOL_WRN("base_url is empty!");
+        return -EINVAL;
+    }
+
+    sol_buffer_init(&buf);
+
+    r = sol_buffer_append_slice(&buf, base_url);
+    SOL_INT_CHECK(r, < 0, r);
 
-    *escaped = sol_buffer_steal(&buf, NULL);
+    if (params && params->params.len) {
+        r = sol_buffer_append_slice(&buf, sol_str_slice_from_str("/?"));
+        SOL_INT_CHECK_GOTO(r, < 0, exit);
+        r = sol_http_encode_params(&buf, SOL_HTTP_PARAM_QUERY_PARAM,
+            params);
+        SOL_INT_CHECK_GOTO(r, < 0, exit);
+    }
 
-end:
+    *uri = sol_buffer_steal(&buf, NULL);
+exit:
     sol_buffer_fini(&buf);
     return r;
 }
+
+static void
+_update_partial_uri(struct sol_str_slice *partial_uri, size_t offset)
+{
+    partial_uri->data = partial_uri->data + offset;
+    partial_uri->len = partial_uri->len - offset;
+}
+
+static int
+_get_scheme(const struct sol_str_slice uri,
+    struct sol_str_slice *next,
+    struct sol_str_slice *scheme)
+{
+    char *sep;
+
+    scheme->data = uri.data;
+
+    sep = memchr(uri.data, ':', uri.len);
+
+    if (!sep) {
+        SOL_WRN("Could not find the scheme seperator (:) at URI:%.*s",
+            SOL_STR_SLICE_PRINT(uri));
+        return -EINVAL;
+    }
+
+    scheme->len = sep - scheme->data;
+
+    if (!scheme->len) {
+        SOL_WRN("Empty scheme. URI: %.*s", SOL_STR_SLICE_PRINT(uri));
+        return -EINVAL;
+    }
+
+    SOL_DBG("URI Scheme: %.*s", SOL_STR_SLICE_PRINT(*scheme));
+    _update_partial_uri(next, scheme->len + 1);
+
+    return 0;
+}
+
+static char *
+_get_first_sep(char *sep1, char *sep2)
+{
+    if (!sep1)
+        return sep2;
+    if (!sep2)
+        return sep1;
+    return sol_min(sep1, sep2);
+}
+
+static int
+_get_authority(const struct sol_str_slice partial_uri,
+    const struct sol_str_slice full_uri,
+    struct sol_str_slice *next,
+    struct sol_str_slice *host,
+    struct sol_str_slice *user,
+    struct sol_str_slice *pass,
+    int *port)
+{
+
+    size_t offset = 2;
+    struct sol_str_slice port_slice = SOL_STR_SLICE_EMPTY;
+    char *sep, *port_sep, *fragment_sep, *path_sep, *query_sep;
+    struct sol_str_slice auth = partial_uri;
+
+    if (!auth.len) {
```

Comment:
you should check if `auth.len > 2`, as you do `strncmp()` later using that amount of bytes... or add 'sol_str_slice_startswith(slice_data, slice_prefix)` that does this in a single step and save errors in the future.